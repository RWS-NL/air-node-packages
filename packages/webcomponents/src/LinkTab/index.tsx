import React, { memo } from 'react';
import { Tab, TabProps } from '@mui/material';

import css from './LinkTab.scss';
import { Link } from 'react-router-dom';

export interface LinkTabProps {
  /** The label to show in the tab */
  label: string;
  /** The URL for this tab to navigate to */
  to: string;
  /**
   * Whether this link should go to an external site or not, using an `<a href=...>` component
   * @default false
   */
  external?: boolean;
  /**
   * Whether this link should open in a new tab or not
   * Only applies when external = true
   * @default false
   */
  openInNewTab?: boolean;

  /** Additional props to pass to each Tab component */
  TabProps?: TabProps;
}

export const LinkTab = memo(({ to, label, external = false, openInNewTab = false, ...props }: LinkTabProps) => {
  if (external) {
    return (
      <Tab
        {...props}
        component='a'
        href={to}
        label={label}
        target={openInNewTab ? '_blank' : '_self'}
        rel={openInNewTab ? 'noreferrer noopener' : undefined}
        classes={{ root: css.linkText, selected: css.selectedTab, textColorInherit: css.linkTextInherit }}
      />
    );
  }

  return (
    <Tab
      {...props}
      component={Link}
      to={to}
      label={label}
      classes={{ root: css.linkText, selected: css.selectedTab, textColorInherit: css.linkTextInherit }}
    />
  );
});
