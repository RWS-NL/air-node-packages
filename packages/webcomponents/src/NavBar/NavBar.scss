@import '../styles/vars';

.tabIndicator {
  display: none;
}

.appBar {
  box-shadow: none;
}
