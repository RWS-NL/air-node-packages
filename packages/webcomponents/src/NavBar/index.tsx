import { AppBar, AppBarProps, Grid, Tabs, TabsProps, Toolbar, ToolbarProps } from '@material-ui/core';
import React, { ChangeEvent, Fragment, memo } from 'react';
import { ActionBar, ActionBarProps } from '../ActionBar';
import { LinkTab, LinkTabProps } from '../LinkTab';
import css from './NavBar.scss';

export interface NavBarProps {
  /** The tabs to render */
  tabs: LinkTabProps[];
  /** The title to show in the ActionBar */
  actionBarTitle: string;
  /**
   *  The currently active tab
   *  Preferably managed in local state
   */
  activeTab: number;
  /** Function to trigger when changing tabs */
  onTabChange(_event: ChangeEvent<object>, newValue: number): void;

  /** Additional props to pass to the AppBar component */
  AppBarProps?: AppBarProps;
  /** Additional props to pass to the Toolbar component */
  ToolbarProps?: ToolbarProps;
  /** Additional props to pass to the Tabs component */
  TabsProps?: TabsProps;
  /** Additional props to pass to the ActionBar component */
  ActionBarProps?: ActionBarProps;
}

export const NavBar = memo(
  ({
    ActionBarProps,
    AppBarProps,
    TabsProps,
    ToolbarProps,
    actionBarTitle,
    activeTab,
    onTabChange,
    tabs
  }: NavBarProps) => (
    <Fragment>
      <AppBar {...AppBarProps} position='static' color='secondary' classes={{ root: css.appBar }}>
        <Toolbar {...ToolbarProps} variant='dense'>
          <Tabs
            {...TabsProps}
            variant='standard'
            value={activeTab}
            onChange={onTabChange}
            classes={{ indicator: css.tabIndicator }}
          >
            {tabs.map((tab, index) => (
              <LinkTab
                key={index}
                label={tab.label}
                to={tab.to}
                external={tab.external}
                data-qa={`link-tab-${index}`}
              />
            ))}
          </Tabs>
        </Toolbar>
      </AppBar>
      <Grid container direction='row' justify='center' alignItems='flex-start'>
        <Grid key={1} item xs={12}>
          <ActionBar {...ActionBarProps} title={actionBarTitle} />
        </Grid>
      </Grid>
    </Fragment>
  )
);
