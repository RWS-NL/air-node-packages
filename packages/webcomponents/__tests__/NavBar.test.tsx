import { Tabs } from '@material-ui/core';
import { LinkTabProps } from '@src/LinkTab';
import { NavBar, NavBarProps } from '@src/NavBar';
import { shallow, ShallowWrapper } from 'enzyme';
import React, { ChangeEvent } from 'react';

let currentValue = 0;
const onTabChange = jest.fn((_event: ChangeEvent<object>, newVal: number) => (currentValue = newVal));
const tabs: LinkTabProps[] = [
  { label: 'root', external: false, to: '/root' },
  { label: 'CGI', external: true, to: 'https://cgi.com' },
  { label: 'Rijkswaterstaat', external: true, to: 'https://rijkswaterstaat.nl', openInNewTab: true }
];

describe('NavBar Tests', function () {
  describe('Component Testing', () => {
    let navBar: ShallowWrapper<NavBarProps>;

    beforeAll(() => {
      navBar = shallow(
        <NavBar actionBarTitle='Title' activeTab={currentValue} onTabChange={onTabChange} tabs={tabs} />
      );
    });

    test('GIVEN currentValue=0 AND I change value THEN currentValue changes', () => {
      const tabs = navBar.find(Tabs);

      tabs.simulate('change', undefined, 1);

      expect(onTabChange).toHaveBeenCalledTimes(1);
      expect(onTabChange).toHaveBeenCalledWith(undefined, 1);
      expect(currentValue).toBe(1);
    });
  });

  describe('Snapshot Testing', () => {
    test('Required Props', () => {
      const navBar = shallow(
        <NavBar actionBarTitle='Title' activeTab={currentValue} onTabChange={onTabChange} tabs={tabs} />
      );
      expect(navBar).toMatchSnapshot();
    });
  });
});
