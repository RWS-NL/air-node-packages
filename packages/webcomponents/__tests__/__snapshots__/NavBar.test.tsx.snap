// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Tests Snapshot Testing Required Props 1`] = `
<Fragment>
  <WithStyles(ForwardRef(AppBar))
    classes={
      Object {
        "root": "appBar",
      }
    }
    color="secondary"
    position="static"
  >
    <WithStyles(ForwardRef(Toolbar))
      variant="dense"
    >
      <WithStyles(ForwardRef(Tabs))
        classes={
          Object {
            "indicator": "tabIndicator",
          }
        }
        onChange={
          [MockFunction] {
            "calls": Array [
              Array [
                undefined,
                1,
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": 1,
              },
            ],
          }
        }
        value={1}
        variant="standard"
      >
        <Memo()
          data-qa="link-tab-0"
          external={false}
          key="0"
          label="root"
          to="/root"
        />
        <Memo()
          data-qa="link-tab-1"
          external={true}
          key="1"
          label="CGI"
          to="https://cgi.com"
        />
        <Memo()
          data-qa="link-tab-2"
          external={true}
          key="2"
          label="Rijkswaterstaat"
          to="https://rijkswaterstaat.nl"
        />
      </WithStyles(ForwardRef(Tabs))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <WithStyles(ForwardRef(Grid))
    alignItems="flex-start"
    container={true}
    direction="row"
    justifyContent="center"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      key="1"
      xs={12}
    >
      <Memo()
        title="Title"
      />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Fragment>
`;
